core/src/main/java/feast/core/model/FeatureGroupInfo.java

```diff
@@ -63,68 +62,75 @@ public FeatureGroupInfo() {
-                    .setServing(servingDataStore)
-                    .build();
-    return FeatureGroupSpec.newBuilder()
-            .setId(id)
-            .addAllTags(TypeConversion.convertTagStringToList(tags))
-            .setDataStores(dataStores)
+        DataStores.newBuilder()
+            .setWarehouse(warehouseDataStore)
+            .setServing(servingDataStore)
             .build();
+    return FeatureGroupSpec.newBuilder()
+        .setId(id)
+        .addAllTags(TypeConversion.convertTagStringToList(tags))
+        .setDataStores(dataStores)
+        .build();
   }
 
   /**
    * Get the feature group detail containing both spec and metadata, associated with this record.
    */
   public FeatureGroupDetail getFeatureGroupDetail() {
     return FeatureGroupDetail.newBuilder()
-            .setSpec(this.getFeatureGroupSpec())
-            .setLastUpdated(TypeConversion.convertTimestamp(this.getLastUpdated()))
-            .build();
+        .setSpec(this.getFeatureGroupSpec())
+        .setLastUpdated(TypeConversion.convertTimestamp(this.getLastUpdated()))
+        .build();
   }
 
-  /**
-   * Checks if this is eq to the other given feature group
-   *
-   * @param otherFeatureGroup
-   * @return boolean
-   */
-  public boolean eq(FeatureGroupInfo otherFeatureGroup) {
-    return otherFeatureGroup.getId() == this.id &&
-            otherFeatureGroup.getTags() == this.getTags() &&
-            otherFeatureGroup.getServingStoreOpts() == this.servingStoreOpts &&
-            getStorageId(otherFeatureGroup.getServingStore()) == getStorageId(this.getServingStore()) &&
-            otherFeatureGroup.getWarehouseStoreOpts() == this.warehouseStoreOpts &&
-            getStorageId(otherFeatureGroup.getWarehouseStore()) == getStorageId(this.getWarehouseStore());
+  public void update(FeatureGroupSpec update) throws IllegalArgumentException {
+    if (!isLegalUpdate(update)) {
+      throw new IllegalArgumentException(
+          "Feature group already exists. Update only allowed for fields: [tags]");
+    }
+    this.tags = String.join(",", update.getTagsList());
+  }
+
+  private boolean isLegalUpdate(FeatureGroupSpec update) {
+    DataStore updatedWarehouseStore =
+        update.getDataStores().hasWarehouse() ? update.getDataStores().getWarehouse() : null;
+    DataStore updatedServingStore =
+        update.getDataStores().hasServing() ? update.getDataStores().getServing() : null;
+    return isStoreEqual(this.warehouseStore, this.warehouseStoreOpts, updatedWarehouseStore)
+        && isStoreEqual(this.servingStore, this.servingStoreOpts, updatedServingStore);
+  }
+
+  private boolean isStoreEqual(StorageInfo oldStore, String oldStoreOpts, DataStore newStore) {
+    return getStorageId(oldStore).equals(newStore == null ? "" : newStore.getId())
+        && oldStoreOpts.equals(
+            newStore == null
+                ? ""
+                : TypeConversion.convertMapToJsonString(newStore.getOptionsMap()));
```

Comment:
Do we really need to have oldStoreOpts here? Can this also be a method on StorageInfo?